### DISCUSSION TITLE
Running custom script - CLI

### DISCUSSION THREAD

**Discussion by john2893** [netbox-community/netbox#13262] [Q&A] [answered]
I can "run script" the custom script via Web. But CLI keeps throwing:
 - ModuleNotFoundError: No module named 'extras.plugins'
 
 - extras module is already installed. 
 - pip installing this separetly doesn't work either: 
  (virtnetbox) [netbox@host bin]$ pip install extras.plugins
     ERROR: Could not find a version that satisfies the requirement extras.plugins (from versions: none) 
     ERROR: No matching distribution found for extras.plugins
     

 
 
`(virtnetbox) [netbox@host bin]$ python3 /usr/local/netbox/scripts/testExporter.py runscript --loglevel debug 

Traceback (most recent call last):
  File "/usr/local/netbox/scripts/testExporter.py", line 8, in <module>
    from netbox_dns.models import View, Zone, Record
  File "/usr/local/sbin/virtnetbox/lib/python3.9/site-packages/netbox_dns/__init__.py", line 1, in <module>
    from extras.plugins import PluginConfig
ModuleNotFoundError: No module named 'extras.plugins'
(virtnetbox) [netbox@host bin]$ `

**Comment by candlerb** [comment]
Custom Scripts and Plugins are two different things.  "PluginConfig" is for plugins only; can you explain why your Custom Script needs to load PluginConfig?  Having said that, I don't know why it works in the web UI.

**Reply by john2893** [reply]
Thank you for replying:

I'm not acutally sure.  I dont _import PluginConfig_ directly in my script, but I see it comes from
```
File "/usr/local/sbin/virtnetbox/lib/python3.9/site-packages/netbox_dns/__init__.py", line 1, in <module>
    from extras.plugins import PluginConfig
    ModuleNotFoundError: No module named 'extras.plugins'
```

I do use "`from netbox_dns.models import View, Zone, Record`" in my script

**Reply by candlerb** [reply]
Ah, so your custom script imports models from a plugin (`from netbox_dns.models import View, Zone, Record`) and the plugin can't initialize itself.

If you can use a plugin this way from a custom script, then it should work both in web and CLI.  If you can make a minimal reproducer (a small complete script, preferably with a minimal plugin) then I think you should be able to raise this as a bug.

**Comment by kkthxbye-code** [comment]
```python3 /usr/local/netbox/scripts/testExporter.py runscript --loglevel debug```

This is not how you run scripts via. the CLI. Check out the docs here:

https://docs.netbox.dev/en/stable/customization/custom-scripts/#via-the-cli

```python3 manage.py runscript [--commit] [--loglevel {debug,info,warning,error,critical}] [--data "<data>"] <module>.<script>```

**Reply by candlerb** [reply]
Ah yes, well spotted :-)

The proviso is that you need be inside an activated virtualenv.  You can avoid this by doing:

```
/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py runscript ...etc
```

**Reply by john2893** [reply]
ohh ok Thank you for putting the path to manage.py in the reply. That was what confused me, I thought thats where I insert my script.

What does <module.script> mean -
Per doc

> module is the name of the python file in the scripts directory without the .py extension 
> <script> is the name of the script class in the <module> to run.

My script has classes-

```
class TestDNExporter(Script):

    class Meta:
        name = "Test Exporter"
        description = "This is a custom script"
        commit_default = True

```
So my CLI will be: Is this correct? 

`python3 /opt/netbox/netbox/manage.py runscript --loglevel debug /usr/local/netbox/scripts/testExporter TestDNExporter `

**Reply by candlerb** [reply, marked as answer]
If the file is `/opt/netbox/netbox/scripts/testExporter.py`, then you would do:

```
/opt/netbox/venv/bin/python /opt/netbox/netbox/manage.py runscript --loglevel debug testExporter.TestDNExporter 
```

By default it won't run a script in an arbitrary directory, but it might be possible with the `--pythonpath` flag.

**Reply by john2893** [reply]
Thank you very much!! 
I see it asking for some module installation now, I'll take care of them. I think this will help me continue my work!!

Really appreciate your help!
